#65: Replace `CRC32` with `BLAKE3`  Hash Function

tareknaser opened the pull request:
### Description
This pull request updates the hashing method for the `ResourceId` struct, replacing `CRC32` with `BLAKE3` hash function.
Related Issue: #65
### Changes 
The changes include:
- Addition of benchmarks for the `ResourceId::compute_bytes` method using the `criterion` crate with the `html_reports` feature enabled. Run `cargo bench` to generate benchmarks in the `/target/criterion` folder
-  Removing collision handling as `BLAKE3` is a cryptographic hash function

> Refer to individual commit messages and bodies for detailed information

tareknaser committed:
feat: add hash function benchmarks

This commit introduces a new benchmark file, benches/hash_benchmark.rs. The file includes the compute_bytes_benchmark function, which has three benchmarks for different use cases:
- compute_bytes_small: Benchmarks with small input data.
- compute_bytes_medium: Evaluates medium-sized input data.
- compute_bytes_large: Measures large input data performance.

criterion is configured as a dev dependency with the html_reports feature for detailed HTML reports in target/criterion/.

Signed-off-by: Tarek <tareknaser360@gmail.com>

tareknaser committed:
feat: change hash function from crc32 to blake3

This commit includes the following changes:

- Update the hashing method in the ResourceId struct from crc32fast to blake3
- Due to the change in hash representation, ResourceId no longer implements Copy but remains Clone.
- Adjust code throughout arklib to accommodate the new hashing method.
- Update tests that utilize ResourceId to reflect the changes.

There will be a follow up commit to remove the collision counter from the code since Blake3 is a cryptographic hash function

Signed-off-by: Tarek <tareknaser360@gmail.com>

tareknaser committed:
fix: remove collision handling

The hash function was updated from crc32 to blake3 in the previous commit. This commit removes references to collisions from the code, as Blake3 is a cryptographic hash function and does not produce collisions.

Signed-off-by: Tarek <tareknaser360@gmail.com>

tareknaser committed:
feat: generate random data for benchmarks

Signed-off-by: Tarek <tareknaser360@gmail.com>

tareknaser committed:
refactor: change storage of BLAKE3 hash to [u8; 32]

This commit updates the storage format of the BLAKE3 hash in the ResourceId struct from a hex-encoded string to a fixed-size array [u8; 32].

Storing the hash as a byte array eliminates the need for hex decoding and encoding operations, potentially improving performance.

Signed-off-by: Tarek <tareknaser360@gmail.com>

kirillt commented:
Let's finish the benchmarks first:
- #74 

After we have index benchmark, we'll be able to judge better about performance penalty of Blake3. Collision tracking, necessary for CRC32, could eat some performance up, too.

kirillt commented on src/id.rs:
```diff
@@ -25,12 +25,12 @@ use crate::{ArklibError, Result};
 impl Display for ResourceId {
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        write!(f, "{}-{}", self.data_size, self.crc32)
+        write!(f, "{}-{:?}", self.data_size, self.blake3)
```
Something like https://crates.io/crates/base64 should be used instead of `{:?}`.

kirillt commented on src/id.rs:
```diff
@@ -40,9 +40,19 @@ impl FromStr for ResourceId {
-        Ok(ResourceId { data_size, crc32 })
+        let s = r.trim_start_matches('[').trim_end_matches(']');
+        let bytes: Vec<&str> = s.split(", ").collect();
+        if bytes.len() != 32 {
+            return Err(ArklibError::Parse);
+        }
+        let mut blake3 = [0u8; 32];
+        for (i, byte_str) in bytes.iter().enumerate() {
+            blake3[i] = byte_str
+                .trim()
+                .parse()
+                .map_err(|_e| ArklibError::Parse)?;
+        }
```
Something like https://crates.io/crates/base64 should be used instead of `{:?}`.

kirillt commented on src/id.rs:
```diff
@@ -25,12 +25,12 @@ use crate::{ArklibError, Result};
 )]
 pub struct ResourceId {
     pub data_size: u64,
```
Actually, we should remove `data_size`, too.

tareknaser committed:
feat: base64 encode blake3 hash

This commit improves readability for ResourceId struct by base64 encoding the blake3 hash for printing and parsing.

It also removes data_size field from ResourceId and removes its references from other parts of the code

Signed-off-by: Tarek <tareknaser360@gmail.com>

tareknaser commented:
This PR should be rewritten when https://github.com/ARK-Builders/arklib/pull/90 is merged here and into `ark-rust` as well. 

I can re rewrite the logic in `ark-rust` directly so that it implements `ResourceIdTrait`

tareknaser commented:
Closing this in favor of https://github.com/ARK-Builders/ark-rust/pull/32
